### PR TITLE
Add documentation for Trees.scala

### PR DESCRIPTION
**Pull request by MasseGuillaume**
This is a WIP, but I would like to have some early feedback.

At a first glance, it looks like a duplication of quasiquote.md. However, it's really useful to see how you can construct a given `scala.meta.Tree` from a scala syntax.

### CODE REVIEW

File: notes/trees.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+Type.Refine           | `A { def f: Int }`
+Type.Existential      | `A forSome { type T }`
+Type.Annotate         | `T @A`
+Type.Lambda           | `[X] => (X, X)`
+Type.[Method]         | ???
+Type.Placeholder      | `T[_ >: Nothing <: Any]`
+Type.Bounds           | `def F[T >: Nothing <: Any] = 1`
+Type.ByName           | `def f(v: => T) = v`
+Type.Repeated         | `def print(v: Any*): Unit`
+Type.Var              | ???
```
Comment (olafurpg) [inline notes/trees.md]:
```scala
ollie-scala@ p"case _: List[t] =>"
res4: Case = Case(Pat.Typed(_, Type.Apply(Type.Name("List"), List(Type.Var(Type.Name("t"))))), None, Term.Block(List()))
```

File: notes/trees.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+Type.ImplicitFunction | `implicit A => B`
+Type.Tuple            | `(A, B)`
+Type.With             | `A with B`
+Type.And              | `A & B`
+Type.Or               | `A \| B`
+Type.Refine           | `A { def f: Int }`
+Type.Existential      | `A forSome { type T }`
+Type.Annotate         | `T @A`
+Type.Lambda           | `[X] => (X, X)`
+Type.[Method]         | ???
```
Comment (olafurpg) [inline notes/trees.md]: n/a without a custom dialect. Can't be written in source
```scala
ollie-scala@ implicit val d = dialects.Scala212.copy(allowMethodTypes = true)
ollie-scala@ "(e: T): T".parse[Type]
res6: Parsed[Type] = Success(Type.Method(List(List(Term.Param(List(), Term.Name("e"), Some(Type.Name("T")), None))), Type.Name("T")))
```

File: notes/trees.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+Term.Eta              | `foo _`
+Term.Repeated         | `foo(x: _*)`
+Term.Param            | `def foo<<(x: Int)>>`
+Term.Interpolate      | `s"|foo ~ $bar ~ ${42}|"`
+Term.Xml              | `<h1>|Hello ~ {bar} ~ {42}|</h1>`
+
+## Types (`meta.Type`)
+
+
+Ref            | Examples
```
Comment (olafurpg) [inline notes/trees.md]: Why is type refs above parse example? In terms it's reversed

File: notes/trees.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Type.Var              | ???
+Type.Param            | `trait A[X]`
+
+[Method]: https://www.scala-lang.org/files/archive/spec/2.12/03-types.html#method-types
+
+
+## Patterns (meta.Pat) and Cases (meta.Case)
+
+```scala
+"""(foo, bar)""".parse[Pat].get.structure
```
Comment (olafurpg) [inline notes/trees.md]: I would use a more obvious pattern like `_: Foo | _: Bar`

**Comment by ShaneDelmore** [comment]
This is pretty wonderful, especially now that you can’t jump to the code in scalameta, only to the protobuf classes.  I don’t know that it needs to hold up this pr but if it was updated with the actual Params to the ADT members (so users can tell how to construct/deconstruct) it would become the canonical reference for users I think.

**Comment by DavidDudson** [comment]
I frequently search in Tree's.scala for the apply/unapply methods for traits and case class constructors to see what parameters are.

Can we add what @ShaneDelmore suggested (parameter lists) and also potentially add the Apply/Unapply for some traits (Although this may deserve going into the tutorial and not this document).

The 2 trait apply/unapply methods are:
- Name
- Lit

**Comment by MasseGuillaume** [comment]
This also might be helpful.

https://github.com/scalacenter/scala-syntax/blob/master/tests/unit/src/test/scala/org/scalafmt/tests/CoverageSuite.scala#L15

**Comment by olafurpg** [comment]
> Can we add what @ShaneDelmore suggested (parameter lists) and also potentially add the Apply/Unapply for some traits (Although this may deserve going into the tutorial and not this document).

We can probably auto-generate that from the sources. I think it's a great idea but I see that complementing the document in this PR. 

@MasseGuillaume saw two rendering issues

screen shot 2018-03-07 at 12 56 56
screen shot 2018-03-07 at 12 57 15

**Comment by olafurpg** [comment]
Closing due to inactivity. Please reopen once the document is more complete @MasseGuillaume ! I'm looking forward to provide this reference to users, I think it's a great idea 😄

**Pull request closed without merging**